src/fu-tool.c

```diff
@@ -752,6 +760,83 @@ fu_util_install (FuUtilPrivate *priv, gchar **values, GError **error)
 	/* success */
 	return TRUE;
 }
+static gboolean
+fu_util_update (FuUtilPrivate *priv, gchar **values, GError **error)
+{
+	g_autoptr(GPtrArray) devices = NULL;
+
+	/* load engine */
+	if (!fu_util_start_engine (priv, error))
+		return FALSE;
+
+	devices = fu_engine_get_devices (priv->engine, error);
+
+	for (guint i = 0; i < devices->len; i++) {
+		FwupdDevice *dev = g_ptr_array_index (devices, i);
+		FwupdRelease *rel;
+		const gchar *device_id;
+		const gchar *remote_id;
+		const gchar *uri_tmp;
+		g_autoptr(GPtrArray) rels = NULL;
+		g_autoptr(GError) error_local = NULL;
+		device_id = fu_device_get_id (dev);
+
+		if (!fu_util_is_interesting_device (dev))
+			continue;
+		/* only show stuff that has metadata available */
+		if (!fwupd_device_has_flag (dev, FWUPD_DEVICE_FLAG_SUPPORTED))
+			continue;
+
+		rels = fu_engine_get_upgrades (priv->engine, device_id, &error_local);
+		if (rels == NULL) {
+			g_printerr ("%s\n", error_local->message);
+			continue;
+		}
+
+		rel = g_ptr_array_index (rels, 0);
+		uri_tmp = fwupd_release_get_uri (rel);
+		remote_id = fwupd_release_get_remote_id (rel);
+		if (remote_id != NULL) {
+			FwupdRemote *remote;
+			g_autoptr(GPtrArray) remotes = NULL;
+			g_autofree gchar *fn = NULL;
+			gchar* install_args[2] = {NULL};
+
+			remotes = fu_engine_get_remotes (priv->engine, error);
+			if (remotes == NULL)
+				return FALSE;
+
+			for (guint j = 0; j < remotes->len; j++) {
+				remote = g_ptr_array_index (remotes, j);
+				if (g_strcmp0 (remote_id, fwupd_remote_get_id (remote)) == 0)
+					break;
+				if (j+1 == remotes->len)
+					remote = NULL;
+			}
+			if (remote == NULL) {
+				g_printerr ("Couldn't find remote %s\n", remote_id);
+				continue;
+			}
+
+			/* local remotes have the firmware already */
+			if (fwupd_remote_get_kind (remote) == FWUPD_REMOTE_KIND_LOCAL) {
+				const gchar *fn_cache = fwupd_remote_get_filename_cache (remote);
+				g_autofree gchar *path = g_path_get_dirname (fn_cache);
+				fn = g_build_filename (path, uri_tmp, NULL);
+				install_args[0] = fn;
```

Comment:
bit eww. Maybe declare argv like `g_auto(GStrv) argv = g_new0 (gchar *, 2)` and then you can do `argv[0] = g_build_filename (path, uri_tmp, NULL);`

Reply:
That's much cleaner, thanks.